src/modules/simulation/gz_bridge/plugins/OpticalFlowSensor.cpp

```diff
@@ -0,0 +1,129 @@
+	gzdbg << "Advertising optical flow data on: " << output_topic << std::endl;
+
+	std::string camera_topic = output_topic;
+	size_t last_segment = camera_topic.rfind("/optical_flow/optical_flow");
+
+	if (last_segment != std::string::npos) {
+		camera_topic = camera_topic.substr(0, last_segment) + "/flow_camera/image";
+	}
+
+	int image_width = 0;
+	int image_height = 0;
+	int update_rate = 0;
+	float hfov = 0;
+
+	auto sensorElem = _sdf.Element()->GetParent()->GetElement("sensor");
+
+	while (sensorElem) {
+		if (sensorElem->Get<std::string>("name") == "flow_camera") {
+			auto cameraElem = sensorElem->GetElement("camera");
+			update_rate = sensorElem->GetElement("update_rate")->Get<int>();
+			hfov = cameraElem->GetElement("horizontal_fov")->Get<double>();
+
+			auto imageElem = cameraElem->GetElement("image");
+			image_width = imageElem->GetElement("width")->Get<int>();
+			image_height = imageElem->GetElement("height")->Get<int>();
+			break;
+		}
+
+		sensorElem = sensorElem->GetNextElement("sensor");
+	}
+
+	gzdbg << "Camera parameters:" << std::endl
+	      << "  image_width: " << image_width << std::endl
+	      << "  image_height: " << image_height << std::endl
+	      << "  update_rate: " << update_rate << std::endl
+	      << "  hfov: " << hfov << std::endl;
+
+	gzdbg << "Subscribing to camera topic for flow: " << camera_topic << std::endl;
+
+	if (!_node.Subscribe(camera_topic, &OpticalFlowSensor::OnImage, this)) {
+		gzerr << "Failed to subscribe to camera topic: " << camera_topic << std::endl;
+		return false;
+	}
+
+	// Assume pinhole camera and 1:1 aspect ratio
+	float focal_length = (image_width / 2.0f) / tan(hfov / 2.0f);
+	_optical_flow = std::make_shared<OpticalFlowOpenCV>(focal_length, focal_length,
+			update_rate, image_width, image_height);
+
+	return true;
+}
+
+void OpticalFlowSensor::OnImage(const gz::msgs::Image &image_msg)
+{
+	if (image_msg.width() == 0 || image_msg.height() == 0) {
+		gzerr << "Invalid image dimensions" << std::endl;
+		return;
+	}
+
+	if (image_msg.pixel_format_type() == gz::msgs::PixelFormatType::RGB_INT8) {
+		cv::Mat temp(image_msg.height(), image_msg.width(), CV_8UC3);
+		std::memcpy(temp.data, image_msg.data().c_str(), image_msg.data().size());
+		cv::cvtColor(temp, _last_image_gray, cv::COLOR_RGB2GRAY);
+
+	} else if (image_msg.pixel_format_type() == gz::msgs::PixelFormatType::L_INT8) {
+		std::memcpy(_last_image_gray.data, image_msg.data().c_str(), image_msg.data().size());
+
+	} else {
+		gzerr << "Unsupported image format" << std::endl;
+		return;
+	}
+
+	uint32_t current_timestamp = (image_msg.header().stamp().sec() * 1000000ULL +
+				      image_msg.header().stamp().nsec() / 1000ULL) & 0xFFFFFFFF;
+
+	if (_last_image_timestamp != 0) {
+		_integration_time_us = (current_timestamp - _last_image_timestamp) & 0xFFFFFFFF;
+	}
+
+	_last_image_timestamp = current_timestamp;
+	_new_image_available = true;
+}
+
+bool OpticalFlowSensor::Update(const std::chrono::steady_clock::duration &_now)
+{
```

Comment:
Small question (might even not be relevent), so we want to calculate optical flow for every simulation step, but not for every image topic received. Wouldn't it make more sense to use the image used at the previous simulation step rather than the last image received?